File: algorithms/cpp/sumofDistancesinTree/sumofDistancesinTree.cpp
// Source: https://leetcode.com/problems/sum-of-distances-in-tree/
// Author: Miao Zhang
// Date:   2021-03-16

class Solution {
public:
    vector<int> sumOfDistancesInTree(int N, vector<vector<int>>& edges) {
        N_ = N;
        res = vector<int>(N, 0);
        count = vector<int>(N, 1);
        graph = vector<vector<int>>(N);
        for (auto& edge: edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        dfs(0, -1);
        dfs2(0, -1);
        return res;
    }

private:
    vector<int> res;
    vector<int> count;
    vector<vector<int>> graph;
    int N_;

    void dfs(int node, int parent) {
        for (int child: graph[node]) {
            if (child != parent) {
                dfs(child, node);
                count[node] += count[child];
                res[node] += res[child] + count[child];
            }
        }
    }

    void dfs2(int node, int parent) {
        for (int child: graph[node]) {
            if (child != parent) {
                res[child] = res[node] - count[child] + N_ - count[child];
                dfs2(child, node);
            }
        }
    }
};

File: algorithms/python/sumofDistancesinTree/sumofDistancesinTree.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/sum-of-distances-in-tree/
# Author: Miao Zhang
# Date:   2021-03-16

###############################################################################
# res[A] = sum(A) + sum(B) + cnt(B)
# res[B] = sum(B) + sum(A) + cnt(A)
# res[A] - res[B] = cnt(B) - cnt(A)
# res[A] = res[B] - cnt(A) + N - cnt(A)
###############################################################################
class Solution:
    def sumOfDistancesInTree(self, N: int, edges: List[List[int]]) -> List[int]:
        graph = collections.defaultdict(set)
        for u, v in edges:
            graph[u].add(v)
            graph[v].add(u)
        count = [1] * N
        res = [0] * N
        
        def dfs(node = 0, parent = None):
            '''
            cnt: Number of nodes with node as root node
            res: Distance of node as root node
            '''
            for child in graph[node]:
                if child != parent:
                    dfs(child, node)
                    count[node] += count[child]
                    res[node] += res[child] + count[child]
        
        def dfs2(node = 0, parent = None):
            for child in graph[node]:
                if child != parent:
                    res[child] = res[node] - count[child] + N - count[child]
                    dfs2(child, node)
        
        dfs()
        dfs2()
        return res
